Single-directory chapter

Julow opened the pull request:
Fixes https://github.com/realworldocaml/book/issues/3052

This PR moves examples from `examples/code` into the corresponding chapter directory (`book/chapter/examples`).

If a chapter has more than one directory in `examples/code/$example_dir` they are moved to `book/chapter/examples/$example_dir`.

Julow commented on book/compiler-frontend/README.md:
````diff
@@ -1090,7 +1090,7 @@ This module defined bindings for the Ncurses library. First, compile the
 files, and then run `ocp-index` in completion mode:
 
-```sh dir=../../examples/code,source-tree=../../examples/code/ffi,skip
+```sh dir=../foreign-function-interface/examples,source-tree=../foreign-function-interface/examples/ffi,skip
````
`compiler-frontend` uses an example from `foreign-function-interface`.

NathanReb replied:
My take on this is that we should duplicate them now thus allowing them to diverge in the future. I think depending on files that aren't within the current directory doesn't cope well with `dune` and it makes maintining the chapters more complex than it should.

Any thought on that @yminsky @avsm ?

avsm replied:
Duplicating is just fine in this case.

Julow replied:
I wonder if duplicating will cause confusion. The text before this block says "Let's refer back to our Ncurses binding example from  [...]" and the block starts with `(cd ffi/ncurses [...]` (this path may have to be updated). Readers should be able to find the source code easily.

NathanReb replied:
I think it'd be okay to duplicate. If we want the code to be strictly identical we can enforce it with a dune `diff` rule for the duplicates in `book/dune` for instance.

No solution is perfect, even the current one as if we rewrite a chapter A without removing the example it's still possible that a later chapter B references out of date examples from A and it wouldn't be enforced by the CI anyway.

Julow commented on book/maps-and-hashtables/README.md:
````diff
@@ -49,7 +49,7 @@ Let's consider an example of how one might use a map in practice. In
 strings. Here's the interface:
 
-```ocaml file=../../examples/code/files-modules-and-programs/freq-fast/counter.mli
+```ocaml file=../files-modules-and-programs/examples/freq-fast/counter.mli
````
Same for `maps-and-hashtables` and `files-modules-and-programs`.

NathanReb commented:
Have you tried removing the gadt examples? If they aren't used anywhere we should simply get rid of them, we can still browse the history if we ever want to bring them back.

NathanReb commented:
Splitting this into one PR per chapter + a wrap up PR would probably ease the review but if it's too much hassle we can do with a single PR!

NathanReb commented:
Nevermind the commit view is good enough!

Julow commented:
The diff is simple, it's just renames and path changes and a single PR will make easy to do other small things.

Removed `gadts`.

The `examples` directory is empty so I removed it, what should I do with the `UNLICENSE` file ?
Should I copy it into every `book/chapter/examples` ? If there were code already in some `book/chapter/examples` should it be an exception ?

NathanReb commented:
Could you fix the duplicate thing so we can merge and I can start working on #3139 ?

Julow commented:
I copied the problematic examples.
Is there some text to rewrite to explain the new scheme to readers ?

NathanReb commented:
Good point! I think the README needs an update but that's true regardless of this PR as it mentions external repos we don't use anymore. Let's open a separate issue for that and merge this as it is!

Julow commented:
This PR removes the `examples/UNLICENSE` file that were applied to every examples I moved.
Should I copy it into every example directories, changing the license of files that already in `book/*/examples`, or to individual directory/file touched by this PR ?
Otherwise, would it be possible to move this file to the root of the project and add the list of file it apply to ?
